cdap-security/src/main/java/co/cask/cdap/security/authorization/DefaultAuthorizationEnforcer.java

```diff
@@ -95,8 +103,18 @@ private void doEnforce(EntityId entity, Principal principal, Set<Action> actions
     if (isAccessingSystemNSAsMasterUser(entity, principal) || isEnforcingOnSamePrincipalId(entity, principal)) {
       return;
     }
-    LOG.debug("Enforcing actions {} on {} for principal {}.", actions, entity, principal);
-    authorizerInstantiator.get().enforce(entity, principal, actions);
+    LOG.trace("==> Enforcing actions {} on {} for principal {}.", actions, entity, principal);
+    // create new stopwatch instance every time enforce is called since the DefaultAuthorizationEnforcer is binded as
+    // singleton we don't want the stopwatch instance to get re-used across multiple calls.
+    StopWatch watch = new StopWatch();
+    watch.start();
+    try {
+      authorizerInstantiator.get().enforce(entity, principal, actions);
+    } finally {
+      watch.stop();
+      LOG.trace("<== Enforced actions {} on {} for principal {}. Time spent in enforcement {} milliseconds.", actions,
```

Comment:
can you automatically WARN, if the duration is >N seconds? N can be like 5 seconds or something.

Reply:
I don't know if we have support for dynamically changing log level. We can always do `if else` but that's not clean and required.
Also if we want some `N` we need to think what that N should be and not have an random value since this is called from various places which will have their own timeout. Also, think about whether its user configurable or not because one `N` will not serve all users.

We can  do it but if the motivation is just making it more configurable and  user friendly I don't think we need to do it at this point.

Reply:
Warning would be nice. You would need a new property in the cConf to make it adjustable. In a similar situation (slow HDFS appends for tx logs) these warnings greatly helped us debug.

Reply:
And if a user reaches out to us for support, and we see such a WARN in the logs, we can just point them to fix their plugin.
Otherwise, its only known at TRACE level, which nobody will have running when they encounter issues.

Reply:
Okay. We will end up introducing a new cConf property in a bug fix release? Is that fine?